Import key origin data through descriptors in importmulti

achow101 opened the pull request:
This PR allows for key origin data as defined by the descriptors document to be imported to the wallet when importing a descriptor using `importmulti`. This allows the `walletprocesspsbt` to include the BIP 32 derivation paths for keys that it is watching that are from a different HD wallet.

In order to make this easier to use, a new field `hdmasterkeyfingerprint` has been added to `getaddressinfo`. Additionally I have removed `hdmasterkeyid` as was planned. I think that this API change is fine since it was going to be removed in 0.18 anyways. `CKeyMetadata` has also been extended to store key origin info to facilitate this.

instagibbs commented:
Nice! strong concept ACK

achow101 commented:
Fixed a bug where the keymetadata wasn't being written to the wallet file. Also added a test for that case.

achow101 commented:
Rebased and also changed the keymeta stuff to add the hd master key id at key generation and when the wallet is first loaded.

luke-jr commented on src/utilstrencodings.cpp:
```diff
@@ -560,6 +560,14 @@ bool ParseHDKeypath(const std::string& keypath_str, std::vector<uint32_t>& keypa
         uint32_t path = 0;
         size_t pos = item.find("'");
+        if (pos == std::string::npos) {
+            // Look for h
```
Why?

achow101 replied:
It's fairly standard to use `h` or `H` as the indicator, not just `'`. In fact, the BIP uses `H`, but as a subscript. `'` is not actually used as the indicator for hardened in the BIP even though that is what everyone uses.

The main reason for allowing this is so that you don't have to do any annoying escaping stuff for the `'` when entering the keypaths on the command line.

achow101 commented:
Rebased

laanwj commented:
Concept ACK, thanks for working on this

achow101 commented:
I will rework this to not require #14019 as soon as I have time (in a few days)

Sjors commented:
+1 for only supporting this with descriptors. Suggest rebasing on #14491, as that currently doesn't save origin info into the wallet.

achow101 commented:
Rebased onto #14491

Sjors commented:
@achow101 can you do a `git commit -a --amend --date "$(date)"` on your own commits so that they appear _after_ the unmerged upstream stuff? It's a bit hard to tell now which commits to review.

Sjors commented:
Seeing some warnings (might be upstream):
<img width="1155" alt="schermafbeelding 2018-11-06 om 10 33 51" src="https://user-images.githubusercontent.com/10217/48055268-7b5cb600-e1af-11e8-85f3-90d2c15bce87.png">

Key origin info from `importmulti` now nicely shows up now in `getaddressinfo`. However, like the upstream PR, it generate invalid psbt's. I tried with a watch-only wallet, manually providing a change address (which isn't used): 

```
src/bitcoin-cli -rpcwallet=... walletcreatefundedpsbt '[{"txid": "...", "vout": 0}]' '[{"...": 0.1}]' 0 '{"includeWatching":  true, "subtractFeeFromOutputs": [0], "changeAddress": "..."}' true

{
  "psbt": "...",
  "fee": 0.00002200,
  "changepos": -1
}

src/bitcoin-cli -rpcwallet=... decodepsbt "..."
error code: -22
error message:
TX decode failed Size of key was not the expected size for the type BIP32 keypath: unspecified iostream_category error
```

Sjors commented:
If we narrow this PR to `WALLET_FLAG_DISABLE_PRIVATE_KEYS` wallets, would that allow you to just reuse `hdseedid` rather than undeprecating and repurposing `hdmasterkeyid `? And if we set `hdseedid` during `importmulti`, would that save you from having to upgrade the wallet version? I.e. get rid of  b0c7bbd4c460fdeb5a91d34f75e96eb115e37598?

achow101 commented:
> If we narrow this PR to WALLET_FLAG_DISABLE_PRIVATE_KEYS wallets, would that allow you to just reuse hdseedid rather than undeprecating and repurposing hdmasterkeyid ?

Yes, but I don't really like that. It would mean that hdseedid has two meanings, depending on whether private keys are disabled. It's definition would not be consistent.

achow101 committed:
Refactor keymetadata writing to a separate method

achow101 committed:
Store key origin info in key metadata

Store the master key fingerprint and derivation path in the
key metadata. hdKeypath is kept to indicate the seed and for
backwards compatibility, but all key derivation path output
uses the key origin info instead of hdKeypath.

achow101 committed:
Add a method to CWallet to write just CKeyMetadata

achow101 committed:
Add WriteHDKeypath function and move *HDKeypath to util/bip32.{h,cpp}

Creates new files util/bip32.h and util/bip32.cpp for containing
BIP 32 stuff.
Moves FormatKeyPath from descriptor.cpp to util/bip32.
Adds a wrapper around it to prepent the 'm' for when just the
BIP 32 style keypath is needed.

achow101 committed:
Implement a function to add KeyOriginInfo to a wallet

achow101 committed:
Import KeyOriginData when importing descriptors

achow101 committed:
Test importing descriptors with key origin information

achow101 commented:
@Sjors I've updated the dates on my commits and fixed the lock issue I believe. I am unable to replicate the creation of invalid PSBTs though. The PSBTs I have created for testing are correct.

Sjors commented:
Isn't the only difference between `hdseedid` and `hdmasterkeyid` that the key with `hdseedid` MAY be present in the wallet (for `!WALLET_FLAG_DISABLE_PRIVATE_KEYS`)? It MAY also be absent for keys generated before `encryptwallet` (which replaces the wallet seed). They both refer to a master key that generated a set of keys in the wallet.

If we migrate to a descriptor based wallet, there would be no difference in treatment between keys with `hdseedid` and keys `hdmasterkeyid` . Or am I missing something? cc @sipa

achow101 commented:
`hdseedid` is the ID of the seed. To get the master key id, the seed needs to be used to generate the master private key, and then the `masterkeyid` generated from that. There's one extra step of abstraction. The hd seed is not the same as the master private key.

Sjors commented:
@achow101 ah, I didn't know about the extra step from seed to master key. That makes sense then.

practicalswift commented on src/util/strencodings.h:
```diff
@@ -202,6 +202,10 @@ bool ConvertBits(const O& outfn, I it, I end) {
 bool ParseHDKeypath(const std::string& keypath_str, std::vector<uint32_t>& keypath);
 
+/** Write HD keypaths as strings */
+std::string WriteHdKeypath(const std::vector<uint32_t> keypath);
```
Should be const reference?

practicalswift commented on src/util/strencodings.cpp:
```diff
@@ -587,6 +587,21 @@ bool ParseHDKeypath(const std::string& keypath_str, std::vector<uint32_t>& keypa
+    return ret;
+}
+
+std::string WriteHdKeypath(const std::vector<uint32_t> keypath)
```
Should be const reference?

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,233 +805,354 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+            if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript " + strRedeemScript + " as this is not a P2SH script.");
+            if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript " + witness_script_hex + " as this is not a P2WSH script.");
+            for (const auto& privkey : import_data.privkeys) {
+                if (import_data.used_keys.count(privkey.first) == 0) {
```
Could use `std::any_of`? :-)

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,233 +805,354 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+            // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
+            if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript " + strRedeemScript + " as this is not a P2SH script.");
+            if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript " + witness_script_hex + " as this is not a P2WSH script.");
+            for (const auto& privkey : import_data.privkeys) {
```
Could use `std::any_of`? :-)

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3536,7 +3536,9 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             ret.pushKV("hdseedid", meta->hd_seed_id.GetHex());
-            ret.pushKV("hdmasterkeyid", meta->hd_seed_id.GetHex());
+            if (!meta->master_key_id.IsNull()) {
+                ret.pushKV("hdmasterkeyid", meta->master_key_id.GetHex());
```
`master_key_id` will often be only filled with the fingerprint rather than the entire ID, perhaps make a note in the help text above or only output the first 4 bytes in that case rather than a lot of zeroes too?

An alternative is to just always store the fingerprint in metadata, never the whole key, because I don't think it's used anywhere else.

achow101 replied:
I've changed this to use the fingerprint by adding `KeyOriginInfo` to `CKeyMetadata`.

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -632,5 +639,52 @@ def run_test (self):
+        # Import pubkeys with key origin info
+        self.log.info("Addresses should have hd keypath and master key id after import with key origin")
+        pub_addr = self.nodes[1].getnewaddress()
+        pub_addr = self.nodes[1].getnewaddress()
```
Why call twice? Is it just to have a non-zero keypath index? Perhaps add comment if deliberate

achow101 replied:
Added a comment.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14021/commits/09995963827b7be1163a50d894490daf182b7a78

A few comments inline

Sjors commented:
Without meaning to expand the number of PR's we need to keep track of, but if this is merged then the `dumpwallet` improvement in #11803 should be updated to look for `master_key_id`. Currently it won't print `master_key_id`, skips origin info and it uses the wrong address type (the latter perhaps unrelated).

achow101 commented:
Rebased

Sjors reviewed (approved):
tACK 39b9de4, modulo depth-4 issue. I don't fully understand the `READWRITE` stuff in 066762f02178b2f39581f903841e75694853d8b7, so someone else should double check that.

Why are you hardcoding a maximum derivation depth of 4? The [BIP32 spec](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#serialization-format) implicitly sets the maximum depth to 256, by having one byte for it. BIP44/49/84 uses 5 levels, though we of course can assume the `coin_type=={0,1}` and get away with 4 levels. But future fancy things like Taproot / MAST might benefit from more levels (e.g. using unique subkeys for various redeem conditions, or a derivation path matching the branch in a Merkle tree).

If it's easy to change the max depth later I don't mind too much.

A bunch of nits inline, maybe better for a followup.

* it would be nice to also expand the test for `dumpwallet`.
* a future backwards compatibility test (#12134) could make sure old versions don't mess with key_origin metadata (I tested manually). Conversely it could test the upgrade functionality (though you can already do that by adding a wallet binary to the test suite).
* for followup, or #15226 if that's merged after this: `importwallet` ignores the origin info that's now added to `dumpwallet`.

Unrelated: it's weird that, according to `dumpwallet`, keys are in random order.

achow101 commented:
> Why are you hardcoding a maximum derivation depth of 4?

I'm not. Where do you see that it does?

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3715,10 +3715,10 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            ret.pushKV("hdkeypath", WriteHdKeypath(meta->key_origin.path));
             ret.pushKV("hdseedid", meta->hd_seed_id.GetHex());
-            ret.pushKV("hdmasterkeyid", meta->hd_seed_id.GetHex());
+            ret.pushKV("hdmasterfingerprint", HexStr(meta->key_origin.fingerprint, meta->key_origin.fingerprint + 4));
```
A `key_origin` `ToString()` method would print both the fingerprint and path. This only needs the fingerprint.

achow101 commented on src/script/sign.h:
```diff
@@ -29,6 +29,25 @@ struct KeyOriginInfo
+
+    void SetNull()
+    {
+        memset(fingerprint, 0, 4);
```
Arrays use less space. This field doesn't need to be dynamic.

instagibbs replied:
I don't think the space savings is worth dealing with IsNull verbosity and the magic number(4).

achow101 replied:
This was something that was already here and I don't think it is necessary to change it.

ajtowns replied:
You could use sizeof(fingerprint) to get rid of the magic number 4? But seems fine as is to me.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -329,6 +337,36 @@ bool CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey,
+            CKeyID master_id = masterKey.key.GetPubKey().GetID();
+            std::copy(master_id.begin(), master_id.begin() + 4, meta.key_origin.fingerprint);
+            if (!ParseHDKeypath(meta.hdKeypath, meta.key_origin.path)) {
+                throw std::runtime_error("Invalid stored hdKeypath");
```
It is atomic per `WriteKeyMetadata`. However I believe that the changes to `CKeyMetadata` are backwards compatible. My understanding of the serialization code is that data at the end of the stream is ignored. So upgraded `CKeyMetadata` should still be open-able in older software.

Sjors replied:
Yes, I also noticed I was able to open (non-blank) 0.18 wallets in 0.17.

sipa replied:
That's correct. If you don't deserialize things at the end of a stream, they just get ignored.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1380,6 +1418,7 @@ CPubKey CWallet::DeriveNewSeed(const CKey& key)
 
     // set the hd keypath to "s" -> Seed, refers the seed to itself
     metadata.hdKeypath     = "s";
```
We need to keep `hdKeypath` forever as `CKeyMetadata` is backwards compatible.

ajtowns replied:
If hdKeypath is being kept forever, does it actually make sense to have the key_origin added to CKeyMetadata? You can calculate key_origin from hdKeypath using ParseHDKeypath() if needed, but everywhere that uses key_origin is converting it back to a string via WriteHdKeypath already as far as I can see? AddKeyOrigin is already filling out hdKeyPath so it would still achieve the point of the PR by the looks?

achow101 replied:
`CWallet::GetKeyOrigin` uses the `key_origin` in `CKeyMetadata`. I strongly prefer storing key origin info in this form instead of having to parse strings every time key origin info is needed to build a PSBT.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -1260,6 +1261,8 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
                 throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
             }
+            pwallet->AddKeyOrigin(pubkey, import_data.key_origins[id]);
+            pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
```
Only new things can be imported, so there shouldn't already be an existing metadata entry.

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -177,6 +177,7 @@ class WalletBatch
     bool WriteTx(const CWalletTx& wtx);
     bool EraseTx(uint256 hash);
 
+    bool WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey, const bool overwrite);
```
Added a comment

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -93,11 +94,13 @@ class CKeyMetadata
     int nVersion;
     int64_t nCreateTime; // 0 means unknown
-    std::string hdKeypath; //optional HD/bip32 keypath
+    std::string hdKeypath; //optional HD/bip32 keypath. Still used to determine whether a key is a seed. ALso kept for backwards compatibility
```
Fixed typo

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -329,6 +337,36 @@ bool CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey,
+
+    for (auto& meta_pair : mapKeyMetadata) {
+        CKeyMetadata& meta = meta_pair.second;
+        if (!meta.hd_seed_id.IsNull() && meta.key_origin.IsNull() && meta.hdKeypath != "s") {
```
Done

achow101 commented on src/script/sign.h:
```diff
@@ -29,6 +29,25 @@ struct KeyOriginInfo
     {
         return std::equal(std::begin(a.fingerprint), std::end(a.fingerprint), std::begin(b.fingerprint)) && a.path == b.path;
     }
+
```
Added a comment

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -855,6 +855,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     //! Load metadata (used by LoadWallet)
     void LoadKeyMetadata(const CKeyID& keyID, const CKeyMetadata &metadata) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     void LoadScriptMetadata(const CScriptID& script_id, const CKeyMetadata &metadata) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    void UpgradeKeyMetadata();
```
This needs a description of what it's doing.

achow101 replied:
Done

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -330,6 +338,36 @@ bool CWallet::WriteKeyMetadata(const CKeyMetadata& meta, const CPubKey& pubkey,
+        CKeyMetadata& meta = meta_pair.second;
+        if (!meta.hd_seed_id.IsNull() && meta.key_origin.IsNull() && meta.hdKeypath != "s") { // If the hdKeypath is "s", that's the seed and it doesn't have a key origin
+            CKey key;
+            GetKey(meta.hd_seed_id, key);
```
Since this is called on load, does the user need to unlock an encrypted wallet?

achow101 replied:
Hmm. I think it will. There is currently no way to get the master key fingerprint without generating it from the hd seed.

instagibbs replied:
I'd like to know what's going to happen here, but we should at least be checking the result here to not calculate bogus data, yes?

achow101 replied:
I think it will write bogus data.

There are a few options to solve this. We could not upgrade the key metadata if we can't calculate the master key id. Or we could use the hd seed id instead if the correct master key id can't be calculated.

I am tending to go towards the second solution. While it technically is not the master key id, this information is really only used for/needed for descriptors and PSBT. Any wallet that already exists does not need this information. If this ends up in PSBT, that's fine since the BIP 32 derivation paths are just information for the signer. In this case, the signer is Bitcoin Core, and it doesn't use the derivation paths. We just need to put something there that won't cause problems for other signers.

sipa replied:
It seems there is an `IsLocked()` at the top now, so you've picked the first solution (not upgrading when locked)? If so, I prefer that.

instagibbs commented on src/wallet/walletdb.h:
```diff
@@ -93,11 +94,13 @@ class CKeyMetadata
     static const int VERSION_BASIC=1;
     static const int VERSION_WITH_HDDATA=10;
-    static const int CURRENT_VERSION=VERSION_WITH_HDDATA;
+    static const int VERSION_WITH_KEY_ORIGIN = 12;
```
why no 11 :(

achow101 replied:
A previous version of this pr used version 11 and I did not want to cause wallets that had tested that pr to become unusable.

Sjors replied:
From my experience older hww branch wallets are already unusable so I wouldn't worry too much about that :-)

Sjors commented:
@achow101 oops, I was confusing the fingerprint (fixed size 4) with path (already a vector). All good then!

utACK 7f84a25, changes since last tACK:
* more comments
* using insert:
<img width="639" alt="schermafbeelding 2019-02-11 om 09 37 44" src="https://user-images.githubusercontent.com/10217/52552515-d1f61380-2de0-11e9-99e4-1e10040becf6.png">

laanwj commented on src/script/sign.h:
```diff
@@ -22,13 +22,32 @@ struct CMutableTransaction;
 struct KeyOriginInfo
 {
-    unsigned char fingerprint[4];
+    unsigned char fingerprint[4]; // First 32 bits of the Hash160 of the public key at the root of the path
```
`//!` please so that it shows up in doxygen

achow101 replied:
Done

laanwj commented on src/script/sign.h:
```diff
@@ -22,13 +22,32 @@ struct CMutableTransaction;
+
+    bool IsNull() const
+    {
+        return path.empty() && fingerprint[0] == 0 && fingerprint[1] == 0 && fingerprint[2] == 0 && fingerprint[3] == 0;
```
:+1: Don't special-case values that can actually occur

Sjors replied:
Can we mandate that `KeyOriginInfo` must always contain a `fingerprint`? So in that case justing checking `path.empty()` is enough.

achow101 replied:
How else then would we determine that a KeyOriginInfo is null?

Sjors replied:
So then it seems we need an explicit "null" for the fingerprint to indicate it's not set.

achow101 replied:
Adding a `null` field would cause a lot of code churn. Instead I've changed `CKeyMetadata` to have a boolean that indicates whether the `key_origin` is valid. Let me know what you think of that; if it's good, I'll squash the commit down.

ajtowns replied:
Looks better to me

laanwj commented on src/util/strencodings.cpp:
```diff
@@ -587,6 +587,21 @@ bool ParseHDKeypath(const std::string& keypath_str, std::vector<uint32_t>& keypa
     return true;
 }
 
+std::string FormatKeyPath(const std::vector<uint32_t>& path)
```
It was never the intent for `strencodings.cpp` to gain so many bitcoin-application-specific utilities, and this application is very narrow. Maybe make a new header with bip32/key specific utils?

achow101 replied:
Moved to `bip32.{h,cpp}`

laanwj commented:
> In order to make this easier to use, the `hdmasterkeyid` field in `getaddressinfo` has been changed to actually give the HD master key id instead of the seed id

This is no longer true with the code change, please update the post (as it will be included in git).

Sjors commented on src/wallet/walletdb.cpp:
```diff
@@ -534,6 +534,9 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         result = pwallet->ReorderTransactions();
 
+    // Upgrade all of the wallet keymetadata to have the hd master key id
+    pwallet->UpgradeKeyMetadata();
```
@ajtowns the new metadata is still readable by older wallets: 
https://github.com/bitcoin/bitcoin/pull/14021#discussion_r255298989

That said, I also have a preference for atomically updating as a general pattern.

achow101 replied:
I've added a catch for the `std::runtime_error`

achow101 committed:
Remove hdmasterkeyid

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3652,7 +3652,7 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "  \"timestamp\" : timestamp,      (number, optional) The creation time of the key if available in seconds since epoch (Jan 1 1970 GMT)\n"
             "  \"hdkeypath\" : \"keypath\"       (string, optional) The HD keypath if the key is HD and available\n"
             "  \"hdseedid\" : \"<hash160>\"      (string, optional) The Hash160 of the HD seed\n"
-            "  \"hdmasterkeyid\" : \"<hash160>\" (string, optional) alias for hdseedid maintained for backwards compatibility. Will be removed in V0.18.\n"
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -855,6 +855,8 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     void LoadKeyMetadata(const CKeyID& keyID, const CKeyMetadata &metadata) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     void LoadScriptMetadata(const CScriptID& script_id, const CKeyMetadata &metadata) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    //! Upgrade stored CKeyMetadata objects to store key origin info as KeyOriginInfo
+    void UpgradeKeyMetadata();
```
Done

achow101 commented:
Addressed comments and updated OP

Sjors commented on src/wallet/walletdb.cpp:
```diff
@@ -529,6 +534,13 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         result = pwallet->ReorderTransactions();
 
+    // Upgrade all of the wallet keymetadata to have the hd master key id
+    try {
```
Nit: add comment that this is not atomic, but that individually updated keymetadata entries don't prevent older wallets from loading.

achow101 replied:
Done

Sjors reviewed (commented):
utACK fa1014c modulo the fingerprint == 0 edge case

Sjors commented:
@achow101 wrote inline:
> Adding a `null` field would cause a lot of code churn. Instead I've changed `CKeyMetadata` to have a boolean that indicates whether the `key_origin` is valid. Let me know what you think of that; if it's good, I'll squash the commit down.

@ajtowns wrote:
> Looks better to me

Agreed, utACK 4908b37 (modulo squash)

achow101 commented:
Squashed

instagibbs commented:
Explicit null key_origin is welcomed change.

I don't think you have addressed the encrypted wallet migration issue with your preferred fix?

achow101 commented:
> I don't think you have addressed the encrypted wallet migration issue with your preferred fix?

Hmm. I thought I pushed that. It must've gotten lost in a rebase somewhere. Pushed it now.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/14021/commits/4486b36762a7ab395d1f01b941d83243af6450d4

Sjors commented:
@instagibbs wrote inline:

> Since this is called on load, does the user need to unlock an encrypted wallet?

@achow101 wrote:
> There are a few options to solve this. We could not upgrade the key metadata if we can't calculate the master key id. Or we could use the hd seed id instead if the correct master key id can't be calculated.

> I am tending to go towards the second solution. While it technically is not the master key id, this information is really only used for/needed for descriptors and PSBT. Any wallet that already exists does not need this information. If this ends up in PSBT, that's fine since the BIP 32 derivation paths are just information for the signer. In this case, the signer is Bitcoin Core, and it doesn't use the derivation paths. We just need to put something there that won't cause problems for other signers.

I don't like this approach; more assumptions than necessary. It seems safer to also run this from `CWallet::Unlock`.

The rest of 4486b36 looks good to me.

instagibbs commented:
@Sjors would it make the first unlock surprisingly long?

Sjors commented:
I'm assuming updating metadata entries is extremely fast, but haven't tested that. You could potentially spit out a warning on first unlock (although that might stress out applications that consume std::out).

achow101 commented:
> I don't like this approach; more assumptions than necessary. It seems safer to also run this from `CWallet::Unlock`.

There's a bit of a problem with doing that. If the wallet is not unlocked but a user decides to use `walletprocesspsbt` with `bip32derivs` set to `true`, then the resulting psbt will have a bogus master key fingerprint. It's writing uninitialized memory to the psbt. This is both a problem in this PR (without the change I just pushed) and in master.

> I'm assuming updating metadata entries is extremely fast

it takes several seconds.

achow101 committed:
Add release notes for importing key origin info change

achow101 commented:
I've reverted that change. In the new push, the key metadata will be upgraded on unlocking the wallet or on loading if the wallet is unencrypted. However, due to the issue I mentioned earlier, I've changed `GetKeyOrigin` to return the `KeyOriginInfo` for keys whose metadata has not been upgraded to just be as if it were an independent key.

Sjors commented:
tACK b384060ff modulo squash.

AppVeyor failure is same as on master.

achow101 commented:
Squashed

Sjors commented:
utACK a8a7103

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -422,6 +465,8 @@ bool CWallet::Unlock(const SecureString& strWalletPassphrase, bool accept_no_key
                 continue; // try another master key
             if (CCryptoKeyStore::Unlock(_vMasterKey, accept_no_keys))
+                // Now that we've unlocked, upgrade the key metadata
+                UpgradeKeyMetadata();
```
This means wallets will iterate through all key metadata every unlock?

Sjors replied:
Perhaps a more efficient approach is to set a wallet flag once all metadata is upgraded?

instagibbs replied:
This is a point in favor of atomicity, then marking as complete.

achow101 replied:
I've added a non-mandatory wallet flag which is set upon the upgrade's completion. This will make sure that the upgrade happens only once.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/14021/commits/8fff73cb092e609505f6a0a14c2fe4d8d80479dd

Sjors commented on src/wallet/wallet.h:
```diff
@@ -135,6 +135,9 @@ enum WalletFlags : uint64_t {
     // wallet flags in the upper section (> 1 << 31) will lead to not opening the wallet if flag is unknown
```
Should this say `upper section (> 0 << 31)` ?

achow101 replied:
Not my problem :p

Sjors commented on src/wallet/wallet.h:
```diff
@@ -135,6 +135,9 @@ enum WalletFlags : uint64_t {
     // unknown wallet flags in the lower section <= (1 << 31) will be tolerated
 
+    // Indicates that the metadata has already been upgraded to contain key origins
+    WALLET_FLAG_KEY_ORIGIN_METADATA = (1ULL << 0),
```
Maybe use `1`? https://github.com/bitcoin/bitcoin/pull/13756/files#diff-12635a58447c65585f51d32b7e04075bR140

achow101 replied:
Done

Sjors reviewed (commented):
utACK 8fff73c modulo Travis; that was smaller change than I expected

I can't reproduce the failing QT test on Travis on macOS 10.14.3, so I restarted the job.

instagibbs commented:
only change is WALLET_FLAG_KEY_ORIGIN_METADATA flag from 0 to 1.

re-utACK https://github.com/bitcoin/bitcoin/pull/14021/commits/1b1c6aa276f24ff85392915c3b557e2343c88252

Sjors commented:
utACK 1b1c6aa

ajtowns commented:
utACK 1b1c6aa276f24ff85392915c3b557e2343c88252

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -1261,6 +1262,8 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             if (!pwallet->GetPubKey(id, temp) && !pwallet->AddWatchOnly(GetScriptForRawPubKey(pubkey), timestamp)) {
                 throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
             }
+            pwallet->AddKeyOrigin(pubkey, import_data.key_origins[id]);
```
Should this be called when `id` is not a member of `import_data.key_origins`? The non-descriptor imports won't have this map entries created.

achow101 replied:
It should check that. I've added the check and a test for this.

sipa commented on doc/release-notes-14021.md:
```diff
@@ -0,0 +1,11 @@
+Miscellaneous RPC Changes
+-------------------------
+- Descriptors with key origin information imported through `importmulti` will have their key origin information stored in the wallet for use with creating PSBTs.
+- If `bip32derivs` of both `walletprocesspsbt` and `walletcreatefundedpsbt` is set to true but the key metadata for a public key has not been updated yet, then that key will have a derivation path as if it were just an independent key (i.e. no derivation path and it's master fingerprint is itself)
```
Typo: `it's` -> `its`

achow101 replied:
Done

sipa reviewed (commented):
utACK, only nits

achow101 commented:
oops, screwed up that most recent push, don't merge yet

MeshCollider commented:
Changes since @sipa's are just nit-fixes, a new test, and a missing brace bugfix :+1:
